### PR TITLE
[10.x] Add APA style title helper

### PR DESCRIPTION
**Pull request by hotmeteor**
This PR provides support for APA style title casing through the `Str::apa($string)` or `str($string)->apa()` helpers.

Style as defined here: https://apastyle.apa.org/style-grammar-guidelines/capitalization/title-case

### CODE REVIEW

**Comment by antonkomarev** [comment]
Isn't it "Title Case" and not APA?

**Comment by jbrooksuk** [comment]
@hotmeteor can you please revert the unrelated styling changes made in `src/Illuminate/Support/Str.php`.

**Comment by hotmeteor** [comment]
@antonkomarev Laravel already has both a `title` and `headline` casing method. This particular method aligns with the APA-defined way of capitalizing specific words in specific positions as a way of titling referenced works.

I'm open to renaming the method, but it's more than just title casing.

@jbrooksuk Done

**Comment by timacdonald** [comment]
I know APA often change their styling rules. What happens when they change their title casing rules?

For example, the heading rules changed from version 6 to version 7 where previously not all titles where "Title Case", but some where "Sentence case".

**Comment by timacdonald** [comment]
An opinionated `Str::taylorCase` 😎

**Comment by hotmeteor** [comment]
@timacdonald Ya I legit don’t care what it’s called, I just used those rules. The point being it’s a nicer way to style titles, especially when referring to works or headlines. 

Totally open to suggestions 😃

File: src/Illuminate/Support/Str.php [hunk 1]
```diff
@@ -110,6 +110,48 @@ public static function afterLast($subject, $search)
         return substr($subject, $position + strlen($search));
     }
 
+    /**
+     * Convert the given string to APA-style case.
+     * Based on the rules provided by the APA:
+     * https://apastyle.apa.org/style-grammar-guidelines/capitalization/title-case
+     *
+     * @param string $value
```
Comment (driesvints) [inline src/Illuminate/Support/Str.php:118]:
```suggestion
     * @param  string  $value
```

**Comment by taylorotwell** [comment]
OK, have made a few changes here 😅 None breaking though...

What I have done:

- Renamed `apa` to `titleCase`
- Added new `properCase` method that is essentially an alias for current `title` method
- Added new `title` argument to `headline`

We should de-document `Str::title` and only document `Str::properCase` and `Str::titleCase`. Then, people will gradually begin to use the new methods for handling these casing scenarios. The current "title" method doesn't make sense and will essentially be "soft deprecated" by removing it from the documentation but we won't remove it from the code for a while.

**Comment by hotmeteor** [comment]
Awesome, thanks @taylorotwell.

Could moving towards this as the "new" `titleCase` be something for 11.x instead?

**Comment by taylorotwell** [comment]
Possibly? I think one improvement to this entire use case could also be the addition of configurable "acronyms". Rails allows this, so that you can do `Str::title('your CSV');` and have it be accurately titled.

**Comment by shaedrich** [comment]
> Isn't it "Title Case" and not APA? I mean APA — American Psychological Association. They are using Title Case, that's all. It's pretty common case, and it should not relate to them.

> @antonkomarev Laravel already has both a `title` and `headline` casing method. This particular method aligns with the APA-defined way of capitalizing specific words in specific positions as a way of titling referenced works.

Even if the result is different, they both result in a title case, one way or the other. One user might prefer the one style, another the other—however, mixing these two (and naming one generic and one specific!) could be quite confusing (especially since [the `Str` helper is already bloated](https://github.com/laravel/framework/discussions/47044#discussioncomment-7034511)).

> I'm open to renaming the method, but it's more than just title casing.

Speaking of being bloated, maybe the `Str` helper is not the right place, because it is quite generic and should be its own helper (`WritingStyle`?) instead (not to quote Taylor but "maybe make it a topic as its use case is quite specific" then) :thinking:

> the heading rules changed from version 6 to version 7 where previously not all titles where "Title Case", but some where "Sentence case".

Does that mean, there would need to be a versioning like with the `ramsey/uuid` package? For example, `(new ApaStyle(version: 7))->apply($text)` (implements `WritingStyle` interface)

> Could moving towards this as the "new" `titleCase` be something for 11.x instead?

> @hotmeteor Possibly? I think one improvement to this entire use case could also be the addition of configurable "acronyms". Rails allows this, so that you can do `Str::title('your CSV');` and have it be accurately titled.

Another possibility than the one mentioned above would be to combine different styling options into their respective methods like this: `Str::title($title, style: WritingStyle::APA)`

As for the rules, this could even made _entirely_ configurable by adding a new configuration file (`config/writing_style.php`) with rules like either
```php
return [
    'rules' => [
        \Apa\TitleCase::class,
        \\ \Apa\SentenceCase::class,
        \\ \ChicagoStyle\SerialComma::class,
    ],
];
```

or

```php
return [
    'title_case' => Env::get('WRITING_STYLE_TITLE_CASE', WritingStyle::APA->value), // 'apa'
    'serial_comma' => Env::get('WRITING_STYLE_SERIAL_COMMA', false),
];
```

Optionally, overwrites could be passed to `Str::style($text, [ 'serial_comma' => true ])` (or `WritingStyle::apply()`)

**Comment by hotmeteor** [comment]
I agree with the sentiment here

**Comment by sakcaysoftware** [comment]
It's only works for English. I wonder if we can make it according to the application language? @hotmeteor @taylorotwell

**Comment by simonhamp** [comment]
Just came to say I think this is a great addition! Thanks for putting in the work @hotmeteor

**Comment by hotmeteor** [comment]
@simonhamp ❤️

@sakcaysoftware I had the same thought. 

I think I might pull this all together into one package:

- different styles
- support different languages

But I have no idea if and how the different styles apply to different languages, will need contributions.

**Comment by sakcaysoftware** [comment]
@hotmeteor  It's great idea. I can help for Turkish language. I think "title style" is enough for different styles you said. It is gone a be very usefull for Blog posts coming from users because not everyone pays attention to that. Basicaly, every words must be start with camelcase except conjunctions. So we should only put language spesific conjuctions array.

**Comment by sakcaysoftware** [comment]
@hotmeteor By conjunctions I mean you defined it as $minorWords in your code

**Comment by hotmeteor** [comment]
@shaedrich You may be interested in shaping this as well

**Comment by sakcaysoftware** [comment]
@hotmeteor you're verrry fast 🚀

**Comment by shaedrich** [comment]
> It's only works for English. I wonder if we can make it according to the application language? @hotmeteor @taylorotwell

> I think I might pull this all together into one package:
> 
> * different styles
> * support different languages
> 
> But I have no idea if and how the different styles apply to different languages, will need contributions.

This circles back to the `config/` file idea from my earlier post

> As for the rules, this could even made _entirely_ configurable by adding a new configuration file (`config/writing_style.php`, similar to [larastan](https://github.com/larastan/larastan/blob/2.x/extension.neon)) with rules like either
> 
> ```
> return [
>     'rules' => [
>         \Apa\TitleCase::class,
>         \\ \Apa\SentenceCase::class,
>         \\ \ChicagoStyle\SerialComma::class,
>     ],
> ];
> ```
> 
> or
> 
> ```
> return [
>     'title_case' => Env::get('WRITING_STYLE_TITLE_CASE', WritingStyle::APA->value), // 'apa'
>     'serial_comma' => Env::get('WRITING_STYLE_SERIAL_COMMA', false),
> ];
> ```
> 
> Optionally, overwrites could be passed to `Str::style($text, [ 'serial_comma' => true ])` (or `WritingStyle::apply()`)

However, since talking about [i18n/l10n](https://laravel.com/docs/localization), the better place might be the `lang/` directory

> I think I might pull this all together into one package
>
> Here's a start of a package... I haven't even run the code 😅 But it should be enough to get it going:
> 
> https://github.com/hotmeteor/titles
>
> @shaedrich You may be interested in shaping this as well

Awesome! I'll definitely check it out later today!

**Comment by cmgmyr** [comment]
@hotmeteor I built a title formatter a while ago at https://github.com/cmgmyr/title-formatter, which is similar to APA, but besides capitalization, it also considers removing additional spaces and a few other special conditions. Feel free to grab any ideas from it!

**Comment by hotmeteor** [comment]
Legend, thanks @cmgmyr!

**Comment by shaedrich** [comment]
> An opinionated `Str::taylorCase` 😎

Sounds custom-taylored to me

**Comment by falconsmilie** [comment]
> @hotmeteor Possibly? I think one improvement to this entire use case could also be the addition of configurable "acronyms". Rails allows this, so that you can do `Str::title('your CSV');` and have it be accurately titled.

APA helper is a fantastic addition to Laravel <3 

The lower casing of acronyms is the only pain point we're experiencing while using it.

**Comment by caendesilva** [comment]
Love this! I had previously created a superset of the title helper, and this is even better!

**Comment by shaedrich** [comment]
> @sakcaysoftware
> 
> Here's a start of a package... I haven't even run the code 😅 But it should be enough to get it going:
> 
> https://github.com/hotmeteor/titles
>
> @shaedrich You may be interested in shaping this as well

Too bad, this never went beyond creation :/

**Comment by shaedrich** [comment]
> Too bad, this never went beyond creation :/

I'm happy to inform you that I talked to @hotmeteor and he informed me that he doesn't think he's pursuing this any further, so will continue the `hotmeteor/titles` project under my form https://github.com/shaedrich/titles/

**Pull request merged**